### ISSUE TITLE
Term Query Loop Block

### ISSUE THREAD

**Issue by tomjn** [WordPress/gutenberg#49094]
## What problem does this address?

Currently you can create a tag cloud, and there's a post term list, but these options are limited, and come in 2 versions ( one for the current post, and one that's just general ). If I need to build anything that deviates from this I'm forced to build a new block.

E.g. if I want to display terms with their term descriptions, or add a graphic next to each tag, sort terms into columns, etc, I can't do that with core blocks. Or even spread terms out in a row the same way a group/row block can.

e.g. creating this type of category list automatically is not possible with the core blocks:

> ## Politics
> News articles about politics
>
> ## Sports
> The latest sports news

## What is your proposed solution?

Instead, A term query block similar to the post query block would be very useful, with term title, term description etc blocks.

This way the existing core blocks could be constructed in any shape form or layout by the users choice without writing code. E.g. using headings for tags with descriptions, custom term featured image blocks in plugins, adding icons, etc

Variants of the block could be registered for each taxonomy too:

<img width="310" alt="Screenshot 2023-03-15 at 12 32 45" src="https://user-images.githubusercontent.com/58855/225309893-92f9c8a4-c143-49d9-948f-0e58a8d0f797.png">

## Proposed Plan

We have a couple of working examples of how these blocks could work ([Terms Query plugin](https://github.com/cr0ybot/term-query) and [a draft PR](https://github.com/WordPress/gutenberg/pull/70682)). However, as this work introduces multiple new blocks, it will be easier to work on and review if the work is split up.

We can use the first container block PR as a base feature branch, and then split the work up as follows:

- [x] Terms Query container block: https://github.com/WordPress/gutenberg/pull/70720
- [x] Terms Template block: https://github.com/WordPress/gutenberg/pull/70747
- [x] Terms Name block (either as a new inner block or a block binding on an existing block): https://github.com/WordPress/gutenberg/pull/71122
- [x] Terms Count block (either as a new inner block or a block binding on an existing block): https://github.com/WordPress/gutenberg/pull/71122

To stabilise the block ready for WP 6.9:
- [x] Modify the term description block to render the preview (or use block bindings): https://github.com/WordPress/gutenberg/pull/71525
- [x] Revamp inspector controls: https://github.com/WordPress/gutenberg/pull/71633
- [x] Combine Order and Order By options: https://github.com/WordPress/gutenberg/pull/71656
- [x] Add Max Terms control: https://github.com/WordPress/gutenberg/pull/71358
- [x] Add variation/pattern picker and multiple layout variations: https://github.com/WordPress/gutenberg/pull/71596
- [x] Add new Term Name block, with option for URL (rather than relying on block bindings): https://github.com/WordPress/gutenberg/issues/71912
- [x] Add Term Count block: https://github.com/WordPress/gutenberg/pull/72218
- [x] Allow exisiting Term Description block to work within Terms Query context: https://github.com/WordPress/gutenberg/pull/72238
- [x] Allow inheriting the terms query from the current template: https://github.com/WordPress/gutenberg/pull/72313
- [x] Add include control: https://github.com/WordPress/gutenberg/pull/72286
- [x] ~Try combining term name and URL in one data binding~: https://github.com/WordPress/gutenberg/pull/71702

**Comment by MaggieCabrera** [comment]
+1 on this. It's a very common pattern for themes that we can't really work around right now. A good example is the section By Topic on Blue Note, a community theme currently in development: https://github.com/WordPress/community-themes/issues/5

**Comment by scruffian** [comment]
The query look actually allows for this already. You can filter the query loop based on a tag or a category:
<img width="280" alt="Screenshot 2023-06-08 at 14 35 36" src="https://github.com/WordPress/gutenberg/assets/275961/45387e06-f21b-474c-8edb-d5b6367ec9a1">

**Comment by tomjn** [comment]
@scruffian you've misunderstood, this issue is not about displaying posts in a term, it's about displaying the terms themselves!

E.g. it would not produce a post list, it could produce a tag list like this:

<img width="493" alt="Screenshot 2023-06-08 at 14 56 20" src="https://github.com/WordPress/gutenberg/assets/58855/1aaaa44e-880e-4c8a-8fd7-a24fe5b6c3e4">

Or more complex items, e.g. a heading with a read more button underneath with a term description

**Comment by scruffian** [comment]
Ah thanks, I thought I must be missing something!

**Comment by huubl** [comment]
+1 this! Are there any updates/decisions made about adding this new block?

**Comment by scruffian** [comment]
> +1 this! Are there any updates/decisions made about adding this new block?

We want it, we just need someone to build it!

**Comment by tomjn** [comment]
Noting that this will need scoping out by an engineer to figure out exactly what needs to be done, it's not as simple as asking for it to be done then someone doing it.

For example, issues for at least these blocks would need to be created:

 - Term Loop
    - category and tag variants with the taxonomy preset
    - term template block
    - no terms block
    - term name/title block
    - term description block

People can probably think of more such as a term slug block but these seem to be the MVP, and each of those will require its own issue and dependencies.

I can also see API changes being desirable re: variations. Afterall if you use a category loop you don't expect a term template, you expect a category template, and a category name/title, making those blocks dependent.

There's also the possibility that because the terms are different entities that it might require taxonomy specific blocks be registered or variants for each taxonomy so that they know which entity type to access and can display the right information.

**_For those versed in agile, this is more like a spike than an implementation ticket, help/contributions needed for specification_**

**Comment by tomjn** [comment]
> +1 this! Are there any updates/decisions made about adding this new block?

This is pretty straight forward to answer. Unless you see PRs mentioned here then there has been no work done to build this. If you can't see any decision making then that means no decision making took place.

If there were a place were decisions were made related to this that didn't filter down it'd be the public weekly editor meetings, but someone would have to raise it on the agenda, and it's highly unlikely the info wouldn't filter over to here. _Unless someone made another issue and that one got traction instead._

**Comment by annezazu** [comment]
Punting to 6.5 for consideration since we're four weeks away from beta 1.

**Comment by annezazu** [comment]
Punted for the same reason as before. This is going to take intentional effort to get done so if anyone is interested for 6.6, please speak up!

**Comment by skorasaurus** [comment]
I'm working with term meta and terms for the first time in several years (or ever?) and came across this issue while reviewing what capabilities exist in Gutenberg in order to find a **block that would display all terms that exist for a particular taxonomy across an entire website**. 

Even after some research about the current state of such a block; I was a bit confused and found this issue which heavily overlaps with https://github.com/WordPress/gutenberg/issues/33302 (This is cross-posted in #33302 as well). 

I include some notes in case anyone else has the same issue or is trying to triage this issue: 

**Categories block** - lists all terms that exist for the taxonomy 'category' across an entire website. 
In the block inserter and list view, the block is called 'Categories List' 
Block markup: `<!-- wp:categories /--> `
Only can select the 'category' taxonomy 

**Post Terms block** - lists the terms that are assigned for a taxonomy (selected in the attribute field) on a particular post or page 
In the block inserter and list view; these blocks are named "Categories" or "tags" 
`<!-- wp:post-terms {"term":"category"} /-->`  or 
`<!-- wp:post-terms {"term":"post_tag"} /--> `

If you register a custom taxonomy, a block variation for the post terms block is registered.
It appears in the block inserter and list view as "Name of your taxonomy(s)"

**Tag cloud** - lists all of the terms that exist for the taxonomy "post_tag" across an entire website. 
in the block inserter and list view, the block is called 'Tag Cloud'; displays it in a cloud like format.  
Only can select the taxonomies 'category' and 'post_tag'
  or  
 
There are a couple existing PRs https://github.com/WordPress/gutenberg/pull/58033 & https://github.com/WordPress/gutenberg/pull/26555 that modifies the categories block to select any taxonomy and displays all of that taxonomy's terms  on the website.

Going forward in the long-term: I think a new flexible block(s) would probably be appropriate as Tom described at https://github.com/WordPress/gutenberg/issues/49094#issue-1625439496

There is a separate block under development - https://github.com/mattyza/terms-list-block but I haven't examined it.

**Comment by aaroncommand** [comment]
@skorasaurus has perfectly described what we and doubtless thousands of other developers need as well - a block that would display all terms which exist for a particular taxonomy. (Ideally with some straightforward way to modify the query being run by that block, so we could filter it by children of a particular term, etc.) It's baffling to me that custom taxonomies are such a fundamental feature of WordPress, people have been asking for some variant of this request for years, and yet resources are being spent to building a real-time multi-user version of Gutenberg or whatever instead of creating this block. For now we are probably just going to hack this up in PHP and punt on the template being block-editor-friendly, since in the old world this is basically just a call to get_terms() that takes two minutes to write.

**Comment by tomjn** [comment]
@aaroncommand he's only in charge of spending his own personal time that's contributed and volunteered here, there's no corporate decision delaying this from being built or product roadmap that punted this.

For this to move forward what's needed is for an engineer to scope out what has to happen so that others know what to do and where to do it. Building this for a single client/customer is easier than making sure this gets built properly for core as tiny engineering decisions can have huge ramifications.

Any support you can show will help here enormously even if it only answers part of the puzzle

**Comment by colorful-tones** [comment]
Hi folks,
We are only one week away from the Beta 1 cut-off date for [WordPress 6.6](https://make.wordpress.org/core/6-6/). This issue hasn’t seen any movement in a while, so we (the editor triage leads of the 6.6 release) have decided to remove it from the [WordPress 6.6 Editor Tasks project board](https://github.com/orgs/WordPress/projects/179/views/1).

**Comment by noisysocks** [comment]
Ran into this while creating a block theme. Definitely a big limitation. There's currently no nice way to create a section like this in a block theme:

<img width="1363" alt="Screenshot 2024-06-04 at 14 41 03" src="https://github.com/WordPress/gutenberg/assets/612155/6fc78138-3ebe-408a-8d98-fb94d8de02d4">

Instead of a new block, could _Categories_ and _Tags_ be an option in the Query Loop block's settings? I suppose maybe not since there'd need to be an accompaniment of related inner blocks e.g. _Taxonomy Title_ and _Taxonomy Count_.

**Comment by ntsekouras** [comment]
>Instead of a new block, could Categories and Tags be an option in the Query Loop block's settings? I suppose maybe not since there'd need to be an accompaniment of related inner blocks e.g. Taxonomy Title and Taxonomy Count.

We cannot use Query Loop. We'd need to update some existing term blocks and add new ones.

**Comment by kauaicreative** [comment]
Came across this implementation of a Term query loop - elegant!
https://brickslabs.com/how-to-list-your-posts-divided-by-categories/
Excited for this! 🔥

**Comment by tomjn** [comment]
@kauaicreative that's using the bricks page builder not Gutenberg

**Comment by kauaicreative** [comment]
> @kauaicreative that's using the bricks page builder not Gutenberg

Yes, bricks. I have not personally used Bricks - but the implementation may offer some inspiration:
1. A single Query Block allows one to query either terns or posts e.g. Choose Terms as query Type and Categories as the taxomomy.
2. It allows for nested query loops which is explained in the referenced tutorial link.

**Comment by fabiankaegy** [comment]
@cr0ybot I'll gladly help review & test your solution in case you plan on bringing them into core 👀 really cool to see some movement here 🤩

**Comment by cr0ybot** [comment]
@fabiankaegy Thanks! That would be really helpful. My approach has basically been to duplicate the Query Loop block and modify it to work with taxonomy terms. I also had to cut out a bunch of stuff for the sake of my own time, like the pattern chooser.

~I got a bit stuck for a while trying to figure out why my bindings from block context weren't displaying properly in the editor until I discovered that custom binding sources still currently display placeholders in the editor!~ Oops, I see there are new editor APIs in 6.7 for rendering custom binding sources in the editor!

**Comment by cr0ybot** [comment]
I have this plugin mostly working, ~though I haven't built any releases yet for easy download/install~ download via releases in the repo. It works in the general sense to list taxonomy terms with customizable child blocks and also can "inherit" the "query" so that you can list terms for a specific post or child terms on a taxonomy term archive.

https://github.com/cr0ybot/term-query

I'm handling the display of term info/meta using block bindings, which is simultaneously cool and somewhat difficult to work with. I've added variations of the heading, paragraph, and button blocks to attach the appropriate block bindings, but obviously these blocks are not scoped to the Taxonomy Terms block only. Not quite sure if there's a better way of handling this sort of thing yet, but I'm definitely not making the same mistake I've made before by adding a bunch of custom blocks for what amounts to some text and links.

**Comment by sarthaknagoshe2002** [comment]
@cr0ybot are you still working on this issue?

**Comment by mikachan** [comment]
👋 That plugin is looking great, @cr0ybot. I like how you've used block bindings rather than creating new blocks. Are you planning on opening a GB PR to include it in Core? I'd be happy to help test and review.

I've also had a go at creating a Terms Query block over in https://github.com/WordPress/gutenberg/pull/70682 (before I saw your plugin!). It's still a work in progress and at the moment I'm mostly testing out implementation ideas. I _did_ create new inner blocks 😅, although only two so far, for the term name and count.

My plan at the moment is to keep tinkering and then eventually open smaller PRs, but I'd be happy to help with your solution instead if you plan to open a PR.

**Comment by sarthaknagoshe2002** [comment]
@mikachan I wonder if he is still working on the plugin. The last commit was quite long ago. Presuming that he is no more working on this issue I too have been working on this since a month. I guess we are gonna end up with a bunch of same stuff 😅

**Comment by mikachan** [comment]
@sarthaknagoshe2002 Oh really?! Well, I'm also happy to help out there 😁 Do you have a PR?

We might be able to separate some parts of the work, e.g. it'd be nice to have the containing Terms Query and Terms Template blocks in place to build on top of, or we could work off a feature branch if that's easier. From working on this over the last few days, my rough plans were:

- Create the containing Terms Query block
- Create Terms Template block
- Add inner blocks, at least for Term Name and Term Count initially. (Or, try using block bindings so we don't need individual blocks.)

Once we have those new blocks, we can then look into expanding the functionality with things like:

- Include/exclude terms controls
- 'Show only child terms' option
- Allow inheriting the terms query from the current template
- More layout variations

**Comment by MaggieCabrera** [comment]
I'm thinking about this block as one like the query loop where patterns will help a lot of users to build their layouts with this block. The main difference with the query loop is that we default inheriting the loop and that makes it really easy for a theme to provide a pattern that is content agnostic. This is way harder for a terms block, where the content can't be inferred by the theme as easily (one might have a specific term type that is very deeply nested, for example). 

What I mean to say is that when we make decisions for this block, we need to think about how it will be used. The way I think about it for this block to be the most useful, it should have a state where a theme can provide a pattern without having to select specifics that are tied to a specific site's content and still be useful, so that the end user only needs to select those specifics to tweak what content to show. I don't know if I'm being clear here?

**Comment by mikachan** [comment]
We're planning on adding an inherit setting similar to the Query block, so that the block inherits the current template's taxonomy context. And then users would still be able to overwrite this with specific settings if necessary. Do you think this would be enough for most patterns?

**Comment by MaggieCabrera** [comment]
> We're planning on adding an inherit setting similar to the Query block, so that the block inherits the current template's taxonomy context. And then users would still be able to overwrite this with specific settings if necessary. Do you think this would be enough for most patterns?

It's hard to say. Most patterns I have seen designed have one to three categories selected and showed (with a related picture) in the home page. I think ideally we want to add the ability to have a featured category image block that we could include in this new block. I understand that's an extra step after the basic foundations are already there, but it's probably what will give more flexibility for layouts given what I've seen designers design for themes so it's good to have in mind as to where we could be walking towards.

I think the inherit option is a good starting point for sure!

**Comment by mikachan** [comment]
Yeah, I've seen images mentioned a bunch of times for categories. I'm thinking that allowing the block to handle inner blocks is a good step forward there, but maybe we could also include some variations that involve images, too. We probably need a new inner block, or this might be a good place to use block bindings on the existing image block.

Thanks, I'll keep these things in mind as we work through this!

**Comment by cr0ybot** [comment]
Hey folks, sorry I've been away from this for a while. I got busy after the initial development of the plugin, and by the time I was satisfied that it was production-ready the core query block (on which I based the term query block) had changed enough that it felt like too much work to take a pass at a core PR.

I also found that I needed a lot more than just a way to query terms, because the display options are limited when compared to posts. I'd love to see term featured images standardized in core the same way featured images for posts are (the only somewhat "standard" term thumbnail is WooCommerce's `thumbnail_id` meta), but since they aren't the core implementation would be limited to the term name, slug, description, link, taxonomy, count, and parent ID, which I've included as block bindings as @mikachan pointed out. I have also included bindings for term meta along with a transform argument so that meta values meant to be an image ID, for instance, could be transformed into a URL for display. I feel like this transform feature could warrant it's own separate feature PR since it shouldn't be limited to bindings only for term meta IMO.

There are also existing core blocks for things like term description which I wasn't able to handle integration for in the plugin, but if the term query block was in code perhaps those other blocks could be updated to integrate properly. Personally though I like the idea of block bindings instead of cluttering the block library with additional highly-specific blocks that may only work as a child to the term query block.

On the topic of inheriting the query, I had some difficulty with that conceptually. Currently all global queries are for posts, not taxonomy terms. It may be possible to detect the taxonomy if on a current term archive, but the usefulness of that is questionable since it displays posts linked to that term. Also, WordPress has no concept of a "taxonomy archive" that just displays terms. Therefore, the block I developed does not inherit by default _unless_ it detects that it is nested within another term query block, in which case inherit is _forced_. I do allow inheriting if the global query is for a term archive, which queries for child terms, or a single post, which fetches the terms attached to the post.

@MaggieCabrera Though maybe not clear from the readme, the plugin handles registered patterns and variations roughly the [same way that the core query block does](https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/extending-the-query-loop-block/#extending-the-block-with-variations), so you don't have to specify a taxonomy within the variation settings itself (and in patterns, the taxonomy is explicitly ignored, for better or worse).

I suppose I should at least make an attempt at a PR with the bones of what I've developed so far.

**Comment by sarthaknagoshe2002** [comment]
> I'd love to see term featured images standardized in core the same way featured images for posts

+1
That would be very useful and it would be cherry on the cake for this block as well!

**Comment by mikachan** [comment]
Thanks for the update, @cr0ybot! 

> It may be possible to detect the taxonomy if on a current term archive

It's currently possible to detect if the Editor context is a taxonomy archive using `getQueryContextFromTemplate`, which will return the taxonomy type and the term. From there, we can build a custom query for listing the terms rather than posts. However, I haven't fully tested this yet, so I'm not sure if it will work for all cases.

It would be great if you're able to open a PR 🙇 We're planning to split up the work in [this draft PR](https://github.com/WordPress/gutenberg/pull/70682) so that it's easier to merge, starting with the Terms Query container block. I think the implementations are similar for the containing blocks, so this may help make landing this in Core smoother.

**Comment by cr0ybot** [comment]
@mikachan I'll take a look at the work-in-progress and see where I can contribute. I'm a little bit confused by how this effort is being split up though, looks like there is currently an [open PR](https://github.com/WordPress/gutenberg/pull/70720) for just the `terms-query` block and the completely separate [draft PR](https://github.com/WordPress/gutenberg/pull/70682) with different commits that includes the same block plus others necessary for it to function. I'm not sure how to contribute unless there is a central "feature" branch/PR that can be branched for sub-features and merged back into the feature branch. Am I thinking about this wrong?

**Comment by mikachan** [comment]
@cr0ybot Thank you! That's right, I've started with the larger draft PR as it's difficult to test any of these blocks in isolation (and also as a proof of concept). But for actually merging these changes, I think it'll be easier to review and iterate if the larger PR is split up, which is why I've started the smaller PR for the Terms Query container block at https://github.com/WordPress/gutenberg/pull/70720.

For building on top of this container block, we could use https://github.com/WordPress/gutenberg/pull/70720 as the feature branch to base further PRs on (in which case I'll rename that PR). I've also set this block as experimental, so I think we'd be fine merging to trunk as an option too. However, I think I'd also prefer working with a feature branch as the base.

**Comment by cr0ybot** [comment]
@mikachan However you want to do it, I'll follow your lead, I have an obvious preference for a feature branch ;]. I just find it a bit harder to work on something like this without the other foundational pieces in place, not sure how we would go about developing the term template block without the term query container block for instance, which makes branching from trunk difficult. Maybe this issue (or a new "tracking issue") should track the separate pieces and link to the PRs instead of that draft PR?

**Comment by mikachan** [comment]
Yes exactly, in that case I think we can try using https://github.com/WordPress/gutenberg/pull/70720 as the feature branch, as that includes the container block we can work from.

> Maybe this issue (or a new "tracking issue") should track the separate pieces and link to the PRs instead of that draft PR?

Yeah good idea, I'll move the checklist over to this issue.

**Comment by MaggieCabrera** [comment]
For reference, the issue about term images is https://github.com/WordPress/gutenberg/issues/64595

**Comment by kmanijak** [comment]
I'm glad to see discussion about images along the taxonomies. I'm writing from WooCommerce perspective where categories can have images and I'd expect them to be displayed.

> Once we have those new blocks, we can then look into expanding the functionality with things like:
> - **Include/exclude terms controls**
> - 'Show only child terms' option
> - Allow inheriting the terms query from the current template
> - **More layout variations**

@mikachan, the two features I highlighted would be crucial from WooCommerce POV. I'll be happy to help with them once[ your PR is merged](https://github.com/WordPress/gutenberg/pull/70720) or should we continue contributing to https://github.com/WordPress/gutenberg/pull/70720?

**Comment by cr0ybot** [comment]
@kmanijak this is the trac ticket associated with taxonomy term images (https://core.trac.wordpress.org/ticket/43835), I think this needs to be done carefully and as a separate effort in core. Since core doesn't currently have a standard taxonomy term thumbnail implementation, I don't think the term query block would be able to support them out of the box initially. My term query block plugin supports things like images via [transform hooks for block bindings](https://github.com/cr0ybot/term-query?tab=readme-ov-file#transformations), which a plugin such as WooCommerce could hook into to provide its own implementation. I have not been able to find time to contribute any code directly to this effort yet outside of what I previously built, but hoping maybe I could contribute some of the block binding stuff for extensibility.

**Comment by mikachan** [comment]
> the two features I highlighted would be crucial from WooCommerce POV. I'll be happy to help with them once your PR is merged or should we continue contributing to https://github.com/WordPress/gutenberg/pull/70720?

@kmanijak Nice, thank you! I think both of these features would be best as follow-ups, separate from https://github.com/WordPress/gutenberg/pull/70720. It'd be great to get this initial block merged, and then we can iterate on both the block itself and any additional features.

**Comment by kmanijak** [comment]
@mikachan, here's the PR with Max Terms control: https://github.com/WordPress/gutenberg/pull/71358.

---

There's a bug with `count` field with hierarchical taxonomies.
- In editor, `count` includes descendants counts
- in frontend, `count` only shows the this term counts

This is example with order by count (descending)

Editor | Frontend
-- | --
<img width="374" height="595" alt="Image" src="https://github.com/user-attachments/assets/0f466f2c-53fa-4a87-84c9-2987546d432b" /> | <img width="444" height="689" alt="Image" src="https://github.com/user-attachments/assets/429402da-1cc3-4534-9b94-110f04c58e4e" />

Interesting though, the order is consistent, just the number differs and honestly I'm wondering which one is "correct". I think it should show accumulated counts with descendants (as in Editor) but I'd love to hear your opinion. Obviously, that would require fixing the actual ordering per count.

Interestingly, @dinhtungdu recently worked on Taxonomy Filters in WooCommerce and I believe that's the strategy he took to always show accumulated count.

**Comment by cr0ybot** [comment]
PR for supporting the term link block binding: #71376

**Comment by sarthaknagoshe2002** [comment]
> Other features that could be follow-up PRs:
>
>  - Modify the term description block to render the preview (or use block bindings)
>  - Add variation picker and multiple layout variations
>  - 'Show only child terms' option
>  - Allow inheriting the terms query from the current template

PR https://github.com/WordPress/gutenberg/pull/71109 adds variation picker, multiple layout, modifies the existing term description block.

However this PR needs some changes due to the newer approach added by the latest PRs.
Once all inner blocks are merged then I'll refine this PR again.
@mikachan although I keep a track but still let me know when you have merged all the inner blocks.

**Comment by mikachan** [comment]
> There's a bug with count field with hierarchical taxonomies.

@kmanijak I think I'm struggling to reproduce this. I've set up three categories: parent category, child category, and grandchild category. I've added an individual post to each category, without assigning the parent categories (e.g. so one post is assigned to just the grandchild category). I see the same counts in both the frontend and the editor:

| Editor | Frontend |
| ---- | ----- |
| <img width="637" height="212" alt="Image" src="https://github.com/user-attachments/assets/2f3c961f-4f01-40e1-a241-ce689d7441b7" /> | <img width="442" height="210" alt="Image" src="https://github.com/user-attachments/assets/dd5be7f6-ace2-43e5-bfbd-7f5c50c1c42b" /> |

But I see your screenshot too! I think we may be able to fix this by adding some additional logic to the data bindings source for the term count.

**Comment by kmanijak** [comment]
> @kmanijak I think I'm struggling to reproduce this. I've set up three categories: parent category, child category, and grandchild category. I've added an individual post to each category, without assigning the parent categories (e.g. so one post is assigned to just the grandchild category). I see the same counts in both the frontend and the editor:

@mikachan, to be fair, my screenshot is with Product Categories taxonomy, apologies I didn't mention it but referenced "Categories" 🙇 which means it may be a problem on Woo side.

**Comment by t-hamano** [comment]
Thanks to everyone who is working on this issue.

Just to be sure, is this block supposed to be shipped with WP 6.9? If so, feel free to add to the 6.9 Editor tasks board.

**Comment by cr0ybot** [comment]
To bring some of the conversation back to this issue, I've opened a new PR that brings the variation picker in line with how the query block handles it, along with adding a pattern picker and a default pattern registration: https://github.com/WordPress/gutenberg/pull/71528

Looking for feedback, in particular about what default patterns and variations should be included.

**Comment by cr0ybot** [comment]
Sorry @sarthaknagoshe2002, I didn't see your comment about a variation picker PR here, there was a bunch of discussion on PR #70720 that I was addressing with #71528 regarding variations and, most importantly, patterns.

I do think modifications to the term description block and additional blocks for term name and count would be welcome improvements but should probably be it's own self-contained PR. I think that these new blocks would probably provide for better default variations as well.

**Comment by sarthaknagoshe2002** [comment]
@cr0ybot Thats completely fine! Don't worry I'll contribute in some other way! Thanks for realizing and apologizing, Certainly not everyone's cup of tea! Cheers to more new blocks by a good community with people like you!

**Comment by cr0ybot** [comment]
@mikachan New PR opened for block pattern picker: #71596

**Comment by cr0ybot** [comment]
Note that the default patterns in #71596 so far are definitely lacking a way to just display the term name and count together as a link. I got close with 2 paragraphs in a row with data bindings but an un-linked term list doesn't seem very useful. A single Term Name block with options for tagName, link, and displaying the count would go a long way here.

**Comment by mikachan** [comment]
> A single Term Name block with options for tagName, link, and displaying the count would go a long way here.

Yes, we may have to create a new Term Name block, as I don't know how we can add the term link otherwise (other than the pattern with the button block). It'd be nice if the button block had a plain style, i.e. like a link block. Or if we could make it possible to bind the term URL to a link format in a paragraph block (maybe this is too complex).

**Comment by cr0ybot** [comment]
Need some eyeballs on PR #71596 which was originally about changing how variations and patterns are handled, and now includes a change in handling hierarchical terms via nesting, inheritance from nesting, post context, and template context, and pattern/query extensibility to bring this block in line with how the query loop block works.

The biggest change is the removal of the `hierarchical` setting so that child terms are instead handled by nesting Terms Query blocks. This makes it possible to handle each level differently, just like the example that @tomjn provided in [this comment](https://github.com/WordPress/gutenberg/issues/49094#issuecomment-1582448857):

> E.g. it would not produce a post list, it could produce a tag list like this:
> 
> <img width="493" height="94" alt="Image" src="https://github.com/user-attachments/assets/d16d6816-0dbb-47b2-ae8a-d73a6c195926" />
> 
> Or more complex items, e.g. a heading with a read more button underneath with a term description

For some background about why hierarchical/nesting changes are needed, see [this comment in the PR](https://github.com/WordPress/gutenberg/pull/71596#issuecomment-3316246058):

> This is going to be more complicated than the query block. I've consolidated into a single `WP_Term_Query` with filterable args, but the results on the front end for hierarchical terms are not quite working the way they seem like they should. You can see the discrepancy right away in the editor vs front end for `perPage=10`:
>
> Editor 	
> 
> <img width="321" height="380" alt="Image" src="https://github.com/user-attachments/assets/1954473e-f229-4dda-8ef9-992bb00233fb" />
>
> Front End
>
> <img width="321" height="380" alt="Image" src="https://github.com/user-attachments/assets/da6abe8e-2598-49c8-b61f-0447658f227a" />
> 
> This is because `WP_Term_Query` is a little weird with limits, and is actually querying ALL terms and [slicing the array](https://github.com/WordPress/WordPress/blob/171b7fb23e873ebe3e8fef7a4ffaf9f30f508345/wp-includes/class-wp-term-query.php#L866-L873). In this case, terms are ordered by name, so the terms that end up in the results array without their own parent also present just get left out.
> 
> Therefore, I admit that separate queries for child terms is actually preferable. However, I'd like to propose a different approach instead of the one currently in use: **nested Terms Query blocks**. We remove the `hierarchical` option entirely and instead allow nested blocks to inherit the current `term_id` of the current block context for use as the `parent` in the query.
> 
> This nested approach feels like it meshes well with the block editor ethos, allowing more control over the presentation of child terms from within the editor. It also follows more closely with the Query block, which does not have a `hierarchical` option even though some post types are hierarchical. The biggest drawback that I can see is that it wouldn't allow for displaying a taxonomy's full hierarchy in one go. However, I'm wondering if this block would really need to be used in that manner, since the Category List block already exists for this use case.

It might also be possible to address issues like #63955 as well if the Query Loops block is updated to support inheriting from the `termId` context. While this could be supported by the current hierarchical approach, I think layout could get messy quick without control over nesting.

Are there any other PRs open for this block that we aren't tracking here? I missed some developments last week that overlapped with the work in progress, particularly in regards to inheritance. We've also been talking about adding a Term Title block and updating the Term Description block to work when nested in a Term Template. These blocks would be ideal for use in the default patterns, block bindings are great but the button block is very one-note when it comes to including links. If I can make a suggestion, I'd love if the Term Title block supported options for showing a link and count.

**Comment by mikachan** [comment]
> Are there any other PRs open for this block that we aren't tracking here?

I've updated the list of issues and PRs in this tracking issue's description, and I don't believe there are currently any other PRs open for this block that we're not tracking here.

> If I can make a suggestion, I'd love if the Term Title block supported options for showing a link and count.

Agreed! I can pick this up in the next few days unless anyone beats me to it (in that case, happy to help review).

**Comment by cr0ybot** [comment]
@mikachan I have split out the nesting & inheritance features from #71596 into #71916 (issue: #71911), which handles inheritance at all levels: nested blocks, taxonomy archive template, and post terms. It also simplifies the `WP_Term_Query` logic and makes it so that the query can be filtered for extensibility. This is a fundamental change that I think makes this block more useful in the long run, and I think we will be working at cross purposes until this is resolved.

**Comment by kmanijak** [comment]
> If I can make a suggestion, I'd love if the Term Title block supported options for showing a link and count.
> Agreed! I can pick this up in the next few days unless anyone beats me to it (in that case, happy to help review).

I'm picking up the inner block(s). My [take on this](https://github.com/WordPress/gutenberg/issues/71912#issuecomment-3338699898) is we need more granular blocks. Hope you agree!

**Comment by kmanijak** [comment]
> I'm picking up the inner block(s). My https://github.com/WordPress/gutenberg/issues/71912#issuecomment-3338699898 is we need more granular blocks. Hope you agree!

@mikachan , sorry for confusion but unfortunately I won't be able to focus on implementing it in upcoming weeks. I'll be involved in discussions though.

**Comment by t-hamano** [comment]
I think there was some discussion somewhere, but is there any PR considering refactoring to move the placeholder from the Terms Template block to the Terms Query block?

Currently, pressing Enter with a placeholder selected doesn't create a new block. I think this issue can be solved by moving the placeholder to the Terms Query block.

<img width="1045" height="444" alt="Image" src="https://github.com/user-attachments/assets/161409f3-d3dd-4b66-879e-11948d77a238" />

**Comment by cr0ybot** [comment]
@t-hamano Yes, sort of. #71596 started out doing this and spiraled out a bit too far. #71916 was meant to address some underlying architectural stuff before circling back to patterns & variations, but now that too may be getting split up further so that's seemingly on hold as well.

**Comment by cr0ybot** [comment]
FYI for others following this issue, @ntsekouras is attempting to trim things down some more in #72107, which mainly involves removing attributes we aren't actively using (yet). Also I've been asked to split the removal of the hierarchical structure out into a separate PR from #71916, will update here when I get a chance to put that together.

As a side note, I fear that the discussion on this effort has fractured a bit, not sure how to remedy that other than trying to provide cross links when I can.

**Comment by t-hamano** [comment]
Perhaps it's a good idea to identify which features are needed, which features could be removed, and whether there are any critical issues to solve in order to stabilize this block and ship v1, and prioritize them.

It's best to keep each task relatively small. [Small PRs are easier to review](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/how-to-get-your-pull-request-reviewed.md#create-the-smallest-reasonable-prs) and allow for quick iteration.

**Comment by t-hamano** [comment]
As I understand it, the first important decision we need to make is whether to introduce the Term Title block and the Term Count block.

- https://github.com/WordPress/gutenberg/issues/71912
- https://github.com/WordPress/gutenberg/issues/71954

**Comment by cr0ybot** [comment]
PR #72121 opened to simplify controls and layout. This mainly removes hierarchical layout structure (precursor to nesting & inheritance) and the `termsToShow` attribute which became redundant when the `termQuery` properties `hierarchical` and `inherit` are used.

**Comment by ntsekouras** [comment]
>I think there was some discussion somewhere, but is there any PR considering refactoring to move the placeholder from the Terms Template block to the Terms Query block?

I can open a PR for this.

IMO next steps are:
1. the above variations issue
2. use inner blocks instead of bindings (see [here](https://github.com/WordPress/gutenberg/pull/70720#discussion_r2366215138))
3. inheritance and extensibility

In every step we still need to see if and how we can simplify things.

**Comment by cr0ybot** [comment]
@ntsekouras I guess I'll close the original patterns PR #71596 then, should I also close #71916 as well and open a new PR for nesting & inheritance instead of resetting/rebasing?

**Comment by cr0ybot** [comment]
There has been some [confusion](https://github.com/WordPress/gutenberg/pull/72121#issuecomment-3377789835) about the `hierarchical` property of `termQuery`, particularly now that the layout itself is no longer nested. The gist is that the REST API doesn't support the `hierarchical` argument, and the `WP_Term_Query` argument of the same name does not do what we are using it for:

> Whether to include terms that have non-empty descendants (even if $hide_empty is set to true). **Default true**.

This means that all term queries performed in the editor will include terms with non-empty descendants, and I made sure the output is the same on the front end by not passing the argument to `WP_Term_Query`.

Perhaps I should have pushed this a little further in #72121, but can we rename it to `showNested`?

While I'm talking about `termQuery` properties, as I'm getting into a minimal nesting & inheritance approach, I'm finding that having separate `parent` and `inherit`* properties may be redundant and can cause some undefined behavior. For instance, what should happen when setting both `parent` to some ID (even 0) and `inherit` to `true` simultaneously? In this block's case, inheritance is based on the `parent` argument that we add to the query. What if we nix the `inherit` property altogether and use `parent="inherit"` for this purpose? So far my experiments with this are working, PR to come soon.

*I know I just added the `inherit` property in #72121, sorry to immediately turn around and consider a different approach.

**Comment by cr0ybot** [comment]
One more note: I've updated the nesting & inheritance issue #71911 to be a sub-tracking issue in order to split up the various inheritance aspects. This is mostly for me to keep my PRs small and focused.

**Comment by cr0ybot** [comment]
PR opened to handle block nesting inheritance only (post term inheritance will be a separate task): #72187

This applies the ideas I mentioned in [my comment](https://github.com/WordPress/gutenberg/issues/49094#issuecomment-3381641181) above.

**Comment by t-hamano** [comment]
I'd like to clarify the remaining tasks to stabilize this block and ship it in 6.9. If there are any breaking changes, they need to be prioritized before stabilizing the block.

**Comment by cr0ybot** [comment]
@t-hamano There are a handful of things I think we need to settle on:

1. The `hierarchical` property of the `termQuery` attribute is confusing, I propose that it be renamed to `showNested`. PR here: #72273
2. The Terms Query block is currently duplicating its own wrapper element in the [PHP render function](https://github.com/WordPress/gutenberg/blob/9688e5b9bc8462571aca3d10826ee9bd2fd8e5d7/packages/block-library/src/terms-query/index.php#L22-L27) because the wrapper is [contained in the `save` function](https://github.com/WordPress/gutenberg/blob/9688e5b9bc8462571aca3d10826ee9bd2fd8e5d7/packages/block-library/src/terms-query/save.js#L6-L10). Currently this block does not actually need to be server rendered since there is no pagination functionality like the Query block has. I think this should be fixed prior to stabilization.
3. PR #72187 for inheriting parent term from a parent Terms Query block is still open; there are some concerns about allowing for this behavior at all. The sub issue for nesting & inheritance is here if we want to focus the discussion elsewhere: #71911
4. I'd like to propose that inheriting terms from the _current post_ context (single template or `postId` context from query loop) be a priority at least if item 3 is not considered important for the stable block.
5. Related to 3 & 4, I had some ideas above for the `inherit` property, particularly for getting rid of it in favor of using the `parent` attribute set to `"inherit"`. That may be slightly short-sighted if inheritance from both a post and term archive are possible, not to mention block context. My main concern is that it could be easy to have an incompatible set of arguments since this block's "inherit" doesn't really work like the Query block (which removes all settings if inheriting). Looking for thoughts here.
6. We don't have a concrete plan for extensibility, and I think we should decide if that is important for the stable block. There is currently no filter for the `WP_Term_Query` args like the [Query block has](https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/extending-the-query-loop-block/) for extending the query (`build_query_vars_from_query_block`). The query logic was simplified in order to make a single query filterable. The `namespace` attribute was also removed, which is part of the Query block's standard method of extension via variations.

**Comment by t-hamano** [comment]
@cr0ybot Thanks for summarizing. So that means we need to address at least 1 and 2 before stabilizing, right?

In order to ship anything new in 6.9, we need to address them **this week**. Also, we need to address anything that could cause compatibility issues after stabilization in advance.

**Comment by t-hamano** [comment]
Please note that if this block is to ship by 6.9, it must be stabilized by **October 17**.

See: https://wordpress.slack.com/archives/C02QB2JS7/p1760337834314219

**Comment by ntsekouras** [comment]
For me we haven't clarified 100% what we want this block to support regarding the nesting and presentation. Do we want to support showing the hierarchy of nested terms (recursively) and have full control of the layout of them? If yes, I don't think we have a good solution for that or that we could find one until Friday. 

Is the above hierarchy nested issue something that we say `Terms List` should handle without full control of the layout, but with css usage and simpler layouts? IMO this could be fine.

The [include](https://github.com/WordPress/gutenberg/issues/71443) functionality IMO is what will make this block used more right from the start, but 🤷 . I'll see to open a PR as soon as I can to test this.




>Currently this block does not actually need to be server rendered since there is no pagination functionality like the Query block has. I think this should be fixed prior to stabilization.

I think this block should remain dynamic, as is a complex one and might server us better with future changes.


>The Terms Query block is currently duplicating its own wrapper element in the [PHP render function](https://github.com/WordPress/gutenberg/blob/9688e5b9bc8462571aca3d10826ee9bd2fd8e5d7/packages/block-library/src/terms-query/index.php#L22-L27) because the wrapper is [contained in the save function](https://github.com/WordPress/gutenberg/blob/9688e5b9bc8462571aca3d10826ee9bd2fd8e5d7/packages/block-library/src/terms-query/save.js#L6-L10).

>The hierarchical property of the termQuery attribute is confusing,

We should look how to simplify things and fix current bugs.


>We don't have a concrete plan for extensibility, and I think we should decide if that is important for the stable block. 

If the first iteration achieves a block that is going to be used with current attributes/settings, I think we can skip it for now. We need to solve the basic issues about what we want to support, simplification and bugs, and focus on those.

**Comment by cr0ybot** [comment]
> The [include](https://github.com/WordPress/gutenberg/issues/71443) functionality IMO is what will make this block used more right from the start, but 🤷 . I'll see to open a PR as soon as I can to test this.

I agree. For those following, @ntsekouras has opened the "include" PR here: #72286

> I think this block should remain dynamic, as is a complex one and might server us better with future changes.

I don't disagree, I've just been pretty consistently overruled when trying to keep something in the codebase that doesn't currently serve a purpose. Still, we need to fix the duplicate wrapper markup bug, which likely means that the render function simply returns `$content`. I'll open that PR in a bit.

Edit: duplicate wrapper bugfix PR: #72296

Edit 2: The above PR has been updated to remove the render function entirely.

**Comment by cr0ybot** [comment]
Noting that several PRs have been opened and merged overnight (from my perspective) that have broken functionality or caused issues elsewhere in the project:

- Term Data block bindings were removed without much discussion: https://github.com/WordPress/gutenberg/pull/72275#issuecomment-3397532291 (needed by #72165)
- Inheritance from custom taxonomy archives was broken by this PR: https://github.com/WordPress/gutenberg/pull/72295#discussion_r2426652310

I know we're in a hurry to stabilize this block but we've got to make sure we're not introducing new issues.

**Comment by ntsekouras** [comment]
>Term Data block bindings were removed without much discussion:

@cr0ybot you're talking about a draft PR for a different block. There is no good way to know that someone intends to use something implemented for Terms Query. So the removal from the code was not a wrong decision.

Having said that, since Navigation block wants it, it's easy to add it back and in fact @getdave pinged me about it and he'll just add it and tailor it for their specific needs.

>Inheritance from custom taxonomy archives was broken by this PR
>I know we're in a hurry to stabilize this block but we've got to make sure we're not introducing new issues.

PRs are reviewed and sometimes something might be missed. Then the bugs should be fixed. It's great that it was caught really fast by you and there is already a fix [here](https://github.com/WordPress/gutenberg/pull/72300).

We're in a hurry to see `if` we'll stabilize the block and for that we need iterations and testing.

**Comment by cr0ybot** [comment]
Fair enough, I'm just a little salty. 🧂🥲

Updates on some in-progress work by @ntsekouras:

- Fix for supporting custom taxonomies when checking the current template's taxonomy has stalled a bit because taxonomy names and terms can't be cleanly derived from the template slug: https://github.com/WordPress/gutenberg/pull/72300
- Feature update for including specific terms, the only downside is not being able to rearrange the selections due to a limitation in the existing controls: https://github.com/WordPress/gutenberg/pull/72286
- Trying again to handle the inherit control similarly to the Query Loop block: https://github.com/WordPress/gutenberg/pull/72313

**Comment by t-hamano** [comment]
> - Feature update for including specific terms, the only downside is not being able to rearrange the selections due to a limitation in the existing controls: https://github.com/WordPress/gutenberg/pull/72286
> - Trying again to handle the inherit control similarly to the Query Loop block: https://github.com/WordPress/gutenberg/pull/72313

These two are enhancements, so if we want to ship the Terms Query block in 6.9, they need to be resolved before Gutenberg 21.9 RC1, the day after tomorrow.

> - Fix for supporting custom taxonomies when checking the current template's taxonomy has stalled a bit because taxonomy names and terms can't be cleanly derived from the template slug: https://github.com/WordPress/gutenberg/pull/72300

This is a bug fix, so it can be addressed in Beta 1 and later.

**Comment by ntsekouras** [comment]
>These two are enhancements, so if we want to ship the Terms Query block in 6.9, they need to be resolved before Gutenberg 21.9 RC1, the day after tomorrow.

I think we can ship both of those for 6.9. I think we should definitely land at least the `inherit` PR. In fact if we land https://github.com/WordPress/gutenberg/pull/72313, we don't need the bug PR (https://github.com/WordPress/gutenberg/pull/72300).

The main problem I'm having is that it's difficult to find folks to provide more opinions about clarifying the main purpose for this block. I've asked in other places too days ago and unfortunately I have no more feedback. @mikachan and @kmanijak who had participated a lot in the discussions are AFK this whole week and that leaves me, @cr0ybot and @t-hamano mostly..

Personally I think we land the two, described are enhancements, this block is in a good place to stabilize, and then address code quality and possible bugs, but 🤷 ..

@t-hamano do you think you could find some time to review/test the above two as well?

**Comment by t-hamano** [comment]
> @t-hamano do you think you could find some time to review/test the above two as well?

I'm a little busy with other tasks today, but I'll try to give you feedback later 👀

**Comment by cr0ybot** [comment]
> The main problem I'm having is that it's difficult to find folks to provide more opinions about clarifying the main purpose for this block. I've asked in other places too days ago and unfortunately I have no more feedback. [@mikachan](https://github.com/mikachan) and [@kmanijak](https://github.com/kmanijak) who had participated a lot in the discussions are AFK this whole week and that leaves me, [@cr0ybot](https://github.com/cr0ybot) and [@t-hamano](https://github.com/t-hamano) mostly..

I definitely have too strong of opinions about this to be one of the only people providing feedback. We do have the slack dev chat happening today, the last one before 21.9 RC1, I've left [a comment](https://make.wordpress.org/core/2025/10/14/dev-chat-agenda-october-15-2025/#comment-47922) asking for this to be part of the agenda.
